### PR TITLE
chore: Remove minor clone points in datadog-grok

### PR DESCRIPTION
**Pull request by blt**
This commit removes some uncessary clones in our datadog-grok implementation. We
are ultimately interested in removing as many allocation points from this
library as possible as the allocations from this crate drown out allocation
analysis from the rest of Vector. Primary culprit are small, temporary
allocations done in onig, not addressed here.

Signed-off-by: Brian L. Troutwine <brian@troutwine.us>

### CODE REVIEW

**Comment by blt** [comment]
Before we can get data from this we'll need to address the test failures, looks like. Vector crashes on startup.

**Comment by blt** [comment]
Alright, I've gone absolutely wild smashing commits and rebasing. I'm going to give b6e62f a shot and see what the results are from that. Less drastic than 7d9555e and also less likely to materially improve allocation patterns, but it might run.

**Comment by blt** [comment]
Can confirm, b6e62f is easier to analyze with heaptrack. It's worth pursuing this, I think, which I believe means getting the tests to pass as next steps.

File: lib/datadog/grok/src/grok.rs [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                r
+            ),
+            Error::GenericCompilationFailure(ref d) => write!(
+                f,
+                "Something unexpected happened during the compilation phase: \"{}\"",
+                d
+            ),
+        }
+    }
+}
```
Comment (blt) [inline lib/datadog/grok/src/grok.rs:293]: Could be redone with thiserror.

File: lib/datadog/grok/src/parse_grok.rs [hunk 2]
```diff
@@ -46,10 +46,7 @@ fn apply_grok_rule(source: &str, grok_rule: &GrokRule, remove_empty: bool) -> Re
     let mut parsed = Value::from(btreemap! {});
 
     if let Some(ref matches) = grok_rule.pattern.match_against(source) {
-        // Extracted fields do not preserve the order(stored in a hashmap)
-        // in which they've been seen in the source expression, which matters for arrays,
-        // so we need to sort them first by keys(grok0, grok1, ...).
-        for (name, value) in matches.iter().sorted() {
+        for (name, value) in matches.iter() {
```
Comment (vladimir-dd) [inline lib/datadog/grok/src/parse_grok.rs:49]: we do need this `.sorted()` call as per the inline comment.

Comment (vladimir-dd) [reply lib/datadog/grok/src/parse_grok.rs:49]: ah, sorry, my bad, didn't notice that we had switched to BTreeMap.

**Review by vladimir-dd — APPROVED** [review]
LGTM, 
P.S. I would just avoid any extra refactoring(e.g. errors) in a copied grok.rs file, since I suspect we might get rid of it completely if we eventually decide to switch to some other regex engine.

**Comment by blt** [comment]
> LGTM, P.S. I would just avoid any extra refactoring(e.g. errors) in a copied grok.rs file, since I suspect we might get rid of it completely if we eventually decide to switch to some other regex engine.

It's sure possible. The only thing grok is doing per se is allowing for definition imports and adding an alias mechanism. That interface is, I think, worth preserving, unless we can upstream the alias notion to some other regex engine and use that directly. It's really a very thin skin, you're right. 

I'll add a comment to that, actually.

**Comment by blt** [comment]
REF #10144

**Pull request merged**